builtin/providers/google/resource_sql_database_instance.go

```diff
@@ -579,6 +603,24 @@ func resourceSqlDatabaseInstanceRead(d *schema.ResourceData, meta interface{}) e
 		_settings["crash_safe_replication"] = settings.CrashSafeReplicationEnabled
 	}
 
+	if v, ok := _settings["disk_autoresize"]; ok && v != nil {
+		if v.(bool) {
+			_settings["disk_autoresize"] = settings.StorageAutoResize
+		}
+	}
+
+	if v, ok := _settings["disk_size"]; ok && v != nil {
+		if v.(int) > 0 && settings.DataDiskSizeGb < int64(v.(int)) {
```

Comment:
Yes, but the problem is the configuration file is demanding the lower value. Taking the following configuration:

```
provider "google" {
  region      = "us-central"
  project     = "redacted"
  credentials = "${file("redacted.json")}"
}

resource "google_sql_database_instance" "instance_master" {
        region = "us-central1"

        settings {
                tier = "db-f1-micro"
                disk_size = 10

                backup_configuration {
                        enabled = true
                        binary_log_enabled = true
                }
        }
}
```

After `apply` the state file will show `"settings.0.disk_size": "10",`. By using the GCP console I manually increase the size to 11 to simulate the auto-resize, without the check in place the next plan will produce the below:
```
~ google_sql_database_instance.instance_master
    settings.0.disk_size: "11" => "10"
```
So the read is getting the correct current value but the configuration is wishing to impose the originally specified 10.

Going on to the `apply` I get the 400 error I was expecting:
```
google_sql_database_instance.instance_master: Refreshing state... (ID: terraform-hex)
google_sql_database_instance.instance_master: Modifying...
  settings.0.disk_size: "11" => "10"
Error applying plan:

1 error(s) occurred:

* google_sql_database_instance.instance_master: Error, failed to update instance terraform-hex: googleapi: Error 400: Invalid instance property., invalidInstanceProperty
```

Side note, the error from the Google API is useless. We first hit this problem when adding a new user and a new schema, I burned a lot of time trying to find the problem with what we were adding before I noticed the shrinking disk.

So the check is in the read so that you don't generate a plan that includes a shrink and it's in the update so that a shrink isn't even attempted.